01-resources-v2/proposal.md

```diff
@@ -0,0 +1,285 @@
+
+* Introduce a "notifier" interface. This is to replace a whole class of resource
+  types that will not be able to fit in to an "artifact" interface.
+
+* Extend the versioned artifacts interface to support deletion of versions,
+  either by an explicit `delete` call or by somehow noticing that versions have
+  disappeared.
+
+
+# Proposal
+
+At this early stage of the RFC, it's easiest for me to just use Elm syntax and
+pretend this is all in a type system.
+
+In reality, each of these functions would be scripts with JSON requests passed
+to them on stdin. I'm going to use the `Bits` type just to represent which calls
+have state on disk to either access (like `put`) or return (like `get`). This
+would normally just be the working directory of the script.
+
+## General Types
+
+```elm
+-- arbitrary configuration
+type alias Config = Dict String Json.Value
+
+-- identifier for space, i.e. {"branch":"foo"}
+type alias Space = Dict String String
+
+-- identifier for version, i.e. {"version":"1.2"}
+type alias Version = Dict String String
+
+-- arbitrary ordered metadata (we may make this fancier in the future)
+type alias Metadata = List (String, String)
+
+-- data on disk
+type alias Bits = ()
+```
+
+## Versioned Artifacts interface
+
+```elm
+spaces  : Config -> Set Space
+check   : Config -> Space -> Maybe Version -> List (Version, Metadata)
+get     : Config -> Space -> Version -> Bits
+put     : Config -> Bits -> (Space, Set Version)
+destroy : Config -> Bits -> Dict Space (Set Version)
+```
+
+## Notifications interface
+
+```elm
+notify : Config -> Notification -> ()
+
+type Notification
+  = BuildStarted
+      { build : BuildMetadata
+      , self : Maybe BuildInput
+      }
+  | BuildFinished
+      { build : BuildMetadata
+      , status : Status
+      , self : Maybe BuildInput
+      }
+
+type alias BuildMetadata =
+  { teamName : String
+  , pipelineName : String
+  , jobName : String
+  , buildName : String
+  , buildID : Int
+  , status : String
+  }
+
+type alias BuildInput =
+  { space : Space
+  , version : Version
+  }
+```
+
+
+# Examples
+
+TODO:
+
+- Pull Requests
+- Feature branches
+- Build matrixes
+- Generating branches (and propagating them downstream)
+- Semver artifacts
+- Fanning out against multiple IaaSes
+- Pool resource?
+- BOSH deploys
+
+
+# Summary of Changes
+
+## Overarching Changes
+
+* Add a `info` script which prints the resource's API version, e.g.
+  `{"version":"2.0"}`. This will start at `2.0`. If `/info` does not exist we'll
+  execute today's resource interface behavior.
+
+* Rather than running `/opt/resource/X`, discover the supported resource APIs by
+  invoking `info`. This allows us to be more flexible in what kinds of resources
+  we can support (versioned artifacts, notifications, ???), and where the
+  scripts may live (`/opt/resource` is very Linux specific).
+
+* Today's resource interface (`/in`, `/out`, `/check`) becomes more specifically
```

Comment:
something that was explicit in the original proposal was that "in" would become "get" and "out" would become "put", and feel like that was a solid rename choice that should carry forward to this proposal.

Reply:
True! I'm planning on that being part of the JSON schema returned by `/info` which I haven't gone into yet. Something like:

```json
{
  "version": "2.0",
  "interfaces": {
    "artifacts": {
      "get": "/opt/resource/get",
      "put": "/opt/resource/put",
      "check": "/opt/resource/check",
      "destroy": "/opt/resource/destroy",
      "spaces": "/opt/resource/spaces"
    }
  }
```

Reply:
oh nice, so you can also build a monolithic binary if you so choose! that'd be great.

Reply:
Yep! We'll want to be careful not to tie ourselves to shell command line parsing, though, so this might take a form similar to the task's [`run` block](https://concourse-ci.org/tasks.html#task-run) instead.